### ISSUE TITLE
Better accessibility with screen readers

### ISSUE THREAD

**Issue by arshaw** [fullcalendar/fullcalendar#2536]
Originally reported on Google Code with ID 2271

```
Inspired by issue 2268. Should research ways to make FullCalendar a better experience
for users with screen readers. Maybe some HTML microformat that does this?
```

Reported by `adamrshaw` on 2014-09-01 22:13:56

Imported with **3** stars.

**Comment by powrsurg** [comment]
Three accessibility issues I can tell you right off the bat:
* On the month view, when you display the "+3 more" when there are too many dates the anchor tag needs an href so that they can tab to the text.
* Continuing from that point, set a tabindex="0" on the close button span on that extended view from the first point so that the user can tab to the close button, and then look for keydown events for the space or enter key.
* Set aria-labels or extra elements with visibly hidden elements to add more text (for example, to make the close button actually read "close").

**Comment by itayst** [comment]
I would like to know what is the status of this issue.
Does full-calendar support today screen readers?

**Comment by barlowm** [comment]
Since I haven't seen any progress on improving the accessibility of this component in meeting with WCAG or Section 508 requirements, does anyone know of any other calendar component that meets accessibility requirements?

**Comment by ish-joshi** [comment]
Here is a way to make the event clicks more accessible. 

Add this option in your fullCalendar configuration.

`eventRender: function (event, element) {
                    element.append("<button class='visuallyhidden' onclick='handleEventClick(" + event + ")' aria-label=' What you want it to say'></button>");
                }`

By adding the invisible button, the screenreader can pick it up. 

Here is the css for .visuallyhidden
`.visuallyhidden {
            border: 0;
            clip: rect(0 0 0 0);
            height: 1px;
            margin: -1px;
            overflow: hidden;
            padding: 0;
            position: absolute;
            width: 1px;
        }`

**Comment by anupshah** [comment]
Hi, I've just recently come across this project and the functionality looks great! Nice one.

On the accessibility side, I wonder if this might help?

If you use aria role=grid, plus various keyboard events to support cursor key navigation, that may give you usability + accessibility enhancements that may also help with screen readers (still needs some additional implementation, mind you)

This has more info: https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/Grid_Role

Hope that helps?

**Comment by wright0776** [comment]
The main accessibility issue that I am having with FullCalendar is the empty table header on the top left of the table in the day view. Could it be filled with a span that is only viewable to screen readers that says "time" or something like that?

**Comment by rawroland** [comment]
Was referred here from the issue #6008 which specifically seeks to add the pressed state to the views section of the toolbar using the `aria-pressed` attribute as described here: https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/button_role.

**Comment by itmaybejj** [comment]
Looking at the current demo, events that are clickable with a mouse (with a link destination) are also clickable with the keyboard. But Click + drag (to move an event) is not available, and days cannot be clicked to switch to the day view.

I don't think events without a link should be in the tabindex. Tabbable items should be interactive, and linkless events are not.

My suggestions for least-effort solutions to add interaction capabilities for keyboards and screen readers (and I may be able to donate some time to offer a patch/pull request; conferring with my team on that...)

1.  Actionable `<a>` elements (like the day links in the month view or the "+ 1 more" links on overflow) must have an `href` for many screen readers to announce they are actionable. Tabindex does let keyboards assign `:focus` to an element, but does _not_ make it so screen readers announce it is clickable. The semantically correct breaking change would be to make these `<button>` elements instead of link elements...buuuuut that would be a breaking change -- the easiest tweak that won't affect existing CSS is to assign the existing links a [role of "button" with a dummy href](https://www.deque.com/blog/accessible-aria-buttons/) (href="#"), and JavaScript handlers for click events that block the jump-to-top behavior -- from `<a class="fc-daygrid-event...">` to `<a class="fc-daygrid-event..." href="#" role="button">`
2. Non-actionable `<a>` elements (events without a link) should be assigned `role="presentation"` so screen readers read them as text instead of links.
3. In the list view, the color dots (.fc-event-graphic) should have hidden alternative text for screen readers if they are meaningful, or the cells should be marked "aria-hidden='true'" if they are not. And yeah; any empty table headers should have hidden text like "time" as mentioned above.
4. For calendars with drag functionality enabled: if an event has a link that goes somewhere, that event works as expected on keyboards and screen readers: it is tabbable and announced as a link. If it does not have a link, it cannot be focused or clicked, even though a mouse can click and drag it. Since something can't have two different keyboard click behaviors (there isn't a "drag") equivalent...well...one common way to address this is to add a visually hidden button after draggable links that becomes visible on `:focus`. Screen readers would read the event and then the button, along the lines of "Link: My event title. Button: Reschedule 'my event title'." Keyboards would tab through link-then-button-then-link-then-button for events with urls, or just button-button-button for events without urls. Perfection! Solves the "people wanting to tab through links" issue. And then for the wizardry: clicking the button would expose a small tooltip that indicates the event can be moved using arrow keys, and each arrow click would shift the event by one day. _(Admittedly getting arrow key hijacking to play nice with screen readers will be challenging; it may be necessary to temporarily set role="application" to disable default screen reader shortcuts, but that is fraught with complications and you would need an aria-live region to warn the user what to do and how to cancel)._ Markup would look something like `<a class="fc-daygrid-event..." role="presentation">Event title</a><button class="show-on-focus" aria-press="false" title="↔" aria-label="Reschedule: event title">Reschedule</button>.` Thaaaaaaaaat would obviously be a ton of work, and if keyboard/screen reader calendar editors have an equivalent alternative interface to schedule events, may not be required by organizations subject to the ADA.

**Comment by gbonker** [comment]
I second @wright0776's point that the main accessibility issue my team is experiencing is the issue of nested tables. Our company recently underwent an accessibility audit from Deque, and this was the main issue with FullCalendar that Deque flagged. 

![image](https://user-images.githubusercontent.com/6987710/109675809-1ed2f100-7b46-11eb-825b-8bb2a6fc8e33.png)
 
Would it be possible to change the HTML that's rendered to only have one main `<table>` tag, with the `<th>`'s containing the days of the week and the `<td>`'s being the time slots? A table nested within a table is confusing to screen reader users. I'm happy to open a separate Github issue if this request would be out of scope or of a different scope than what the original poster of this issue had in mind.

**Comment by sreynen** [comment]
Is this issue a bug report or a feature request? Could it be re-written using one of those templates? I ask because this seems like it's actually multiple issues, none of which are moving forward because they're all bundled together, and I suspect trying to use the issue templates would make that clearer.

**Comment by dungeonjumper** [comment]
@sreynen The problem with this issue is that it's not actionable at all. It's simply asking for _better_ accessibility with screen readers. Considering that this calendar currently works poorly with screen readers, there are too many things that need to be done that can be tracked in a single issue. The even bigger problem is that users are opening specific, actionable issues that all get sent here and closed, meaning there is little to no movement on the a11y front with this project. We can't lump all a11y-related issues into one non-specific issue because the lack of clarity is hindering progress.

**Comment by sreynen** [comment]
@dungeonjumper I agree. That's why I was suggesting rewriting it based on a template, because the templates are designed to make issues actionable. There are other actionable accessibility issues, but they have been closed in favor of this non-actionable issue.

**Comment by arshaw** [comment]
I am deeply sorry for taking so long to address these accessibility concerns. I've broken things down into separate, actionable tickets. They are divided into two milestone:

- [a11y-soon](https://github.com/fullcalendar/fullcalendar/milestone/46) - low-hanging fruit that will be addressed in the next minor release
- [a11y-future](https://github.com/fullcalendar/fullcalendar/milestone/46) - tasks that require significant refactoring. probably available in the next major version

If particular tickets interest you, please don't forget to **give a** 👍  **to vote** ([?](https://github.com/blog/2119-add-reactions-to-pull-requests-issues-and-comments)), press the 🔈**Subscribe** button to receive updates.

I am closing this master ticket in favor of the individual tickets.

Special thanks to @powrsurg and @itmaybejj for the great advice.
